File: compiler/java/com/google/dart/compiler/type/Type.java
// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

package com.google.dart.compiler.type;

import com.google.dart.compiler.resolver.Element;

import java.util.List;

/**
 * Common supertype of all types.
 */
public interface Type {
  /**
   * Performs the substitution [arguments[i]/parameters[i]]this.
   * The notation is known from this lambda calculus rule:
   * (lambda x.e0)e1 -> [e1/x]e0.
   * <p>See {@link TypeVariable} for a motivation for this method.
   */
  Type subst(List<Type> arguments, List<Type> parameters);

  Element getElement();

  TypeKind getKind();
  
  /**
   * @return the {@link TypeQuality}, not <code>null</code>.
   */
  TypeQuality getQuality();
}

File: pkg/analysis_server/lib/src/utilities/extensions/type.dart
// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'package:analysis_server/src/utilities/extensions/element.dart';
import 'package:analyzer/dart/element/type.dart';

/// Extensions for [DartType]s.
extension DartTypeExtensions on DartType {
  /// Return `true` if this type represents the class `Iterable` from
  /// `dart:core`.
  bool get isDartCoreIterable =>
      this is InterfaceType &&
      (this as InterfaceType).element.isDartCoreIterable;
}
